src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java

```diff
@@ -25,94 +53,1321 @@
+    
+    // TODO multiple page data classes inheriting this for each view type, 
+    // rather than an enum determining behaviour in many methods
+    ViewType viewType;
+    enum ViewType {
+        QUESTION, GIVER_QUESTION_RECIPIENT, RECIPIENT_QUESTION_GIVER;
+        
+        public String toString() {
+            return name().toLowerCase().replaceAll("_", "-");
+        }
+        
+        public boolean isFirstGroupedByGiver() {
+            return this == GIVER_QUESTION_RECIPIENT;
+        }
+    }
+    
+    
     public InstructorFeedbackResultsPageData(AccountAttributes account) {
         super(account);
         startIndex = -1;
     }
+    
+    public void initForViewByQuestion(InstructorAttributes instructor, 
+                                      String selectedSection, String showStats, 
+                                      String groupByTeam) {
+        viewType = ViewType.QUESTION;
+        
+        this.instructor = instructor;
+        this.selectedSection = selectedSection;
+        this.showStats = showStats;
+        this.groupByTeam = groupByTeam;
+        
+        Map<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>> questionToResponseMap = bundle.getQuestionResponseMap();
+        questionPanels = new ArrayList<InstructorResultsQuestionTable>();
+        
+        for (Map.Entry<FeedbackQuestionAttributes, List<FeedbackResponseAttributes>> entry : questionToResponseMap.entrySet()) {
+            FeedbackQuestionAttributes question = entry.getKey();
+            List<FeedbackResponseAttributes> responses = entry.getValue();
+            
+            questionPanels.add(buildQuestionTableAndResponseRows(question, responses, ""));
+        }
+        
+    }
+    
+    /**
+     * Creates {@code InstructorFeedbackResultsSectionPanel}s for sectionPanels.
+     * 
+     * Iterates through the responses and creates panels and questions for them. Keeps track 
+     * of missing sections, teams and participants who do not have responses 
+     * and create panels for these missing sections, teams and participants.
+     * 
+     * TODO: simplify the logic in this method
+     */
+    public void initForViewByGiverQuestionRecipient(InstructorAttributes instructor, 
+                                    String selectedSection, String showStats, 
+                                    String groupByTeam) {
+        this.instructor = instructor;
+        this.selectedSection = selectedSection;
+        this.showStats = showStats;
+        this.groupByTeam = groupByTeam;
+        
+        if (!bundle.isComplete) {
+            // results page to be loaded by ajax instead 
+            buildSectionPanelsForForAjaxLoading(sections);
+            return;
+        }
+        
+        // Note that if the page needs to load by ajax, then responses will be empty too,
+        // therefore the check whether the bundle needs to come before this
+        if (bundle.responses.isEmpty()) {
```

Comment:
Is it possible for responses to be not empty and bundle to be incomplete at the same time?
If it is, the comment should be amended to show that the responses is not guaranteed to be empty.
If it's not, would it be good to nest the ajax check in the empty check?